### DISCUSSION TITLE
Web Scrollbar Behavior and Headers

### DISCUSSION THREAD

**Discussion by leggomuhgreggo** [react-navigation/react-navigation#10597] [Ideas]
### Standard Web Scrollbar Behavior

Here's an example of the standard scrollbar behavior on web:

![image](https://user-images.githubusercontent.com/2213636/168966308-e995fee1-6f19-4bf1-a9d7-b7825e9b18eb.png)

You can see that:
- the scrollbar goes the full height of the window
- the document content is separate.

### React Navigation Scrollbar Behavior

Here're a couple example cases demonstrating some issues with the current approach for web.

![image](https://user-images.githubusercontent.com/2213636/168965758-7a2b5a66-0122-49a4-8dcb-8df82c881fb2.png)

As you can see the nested scrolling context isn't fully coherent w/ the Header presentation -- and the result clashes a bit with the standard web scrolling idioms. 

#### Packages
I haven't made an exhaustive investigation across all navigators + settings, but I believe this behavior is happening w/ Stack and Native Stack.

#### Related Discussions

I searched a good bit and couldn't find anything specifically related to this issue. Regrets if I've overlooked something.

## Possible Solutions

Normally, I'd say the solution would be to have the Header render with a position relative to the inner boundary of the overflow container -- but, in this case I'm not sure that's possible, since the Scrollview would be user-defined within the Screen component, at a lower nesting level than React Navigation code that manages Headers.

1. Expose new option to disable absoluteFill of Screen component markup, in favor of a custom ~root-level scroll container
2. Expose some "context-connected" HeaderComponent to compose into custom ScrollViews, perhaps leveraging stickyHeaderIndices [[codesandbox](https://codesandbox.io/s/react-native-web-stickyheaderindices-ywu2n0?file=/src/App.js)]
3. Absolute position the header, with offset so that it doesn't obscure the nested contexts scrollbar (hacky)
4. something else?

### Discussion

Option 1 would perhaps be the most "pure" with respect to web idioms, but it would seem to mean only having one scroll component for the entire app.

Option 2 (or something like it) seems like the most preferable of the three directions, but I could see that inviting other potential complexities.

Option 3 this is probably an example of what we don't want to do. But maybe it's less fragile than Im imagining. 

I suspect there are other approaches than Im not considering. Would love to hear others' thoughts on the subject.

Thanks!

---

_Note: This issue is most easily recognized if you update your scrollbar behavior to "Always"_

![image](https://user-images.githubusercontent.com/2213636/168965293-1024096e-dcc0-4289-baee-7df448353f7e.png)

**Comment by 6TELOIV** [comment]
Found this in 2025, still kinda an "issue". Beyond the cosmetics of the position of the scrollbar, this also has implications for mobile website users. On websites where the body scrolls, Safari and Chrome on iOS and Android, respectively, will hide the navbar in response to a user scrolling down a page, and show it again when they scroll back up. Currently, with react navigation, this doesn't seem possible, as the body is fixed to fill the screen, and the header/tabs are placed before/after the screen.

I think the best way to do this would be with sticky positioning on the header/tabs, and allow the main screen to be whatever height it needs to be, at least on the web. In native apps, the current behavior seems to be the way to go. Also, not all apps follow this structure, so maybe more nuanced solutions are needed.

**Comment by Bram-dc** [comment]
To gain more control over the header, I wrapped the entire layout in a scroll view, allowing the header to scroll with the screen content. I would love to be able to attach the header inside screenLayout instead.

```layout: (props) => <ScrollView {...props} contentContainerStyle={{ flex: 1 }} />,```
